cat-code-arduino/CR-BW0-B00-M00/CR-BW0-B00-M00.ino

```diff
@@ -0,0 +1,371 @@
+
+const char strEnder = 0xfa;
+const char delimiter = 0xff;
+const char mouseClickLeft = 0xfe;
+const char mouseClickRight = 0xfc;
+const char mouseClickMiddle = 0xfd;
+
+byte eL;
+
+int k;  //loop variable for subfunctions
+
+const char ctrl = 0x80;
+byte ctrl_pressed = 255;
+const char shift = 0x81;
+byte shift_pressed = 255;
+
+
+
+#define AInput 1      //pins for encoder
+#define BInput 0
+byte lastState = 0;   // variables for encoder 
+byte steps = 0;
+int  cw = 0;
+byte AState = 0;
+byte BState = 0;
+byte State = 0;
+
+
+void setup() {
+  pinMode(safetyPin,INPUT);       //running the SAFETY-loop as long pin 21 is disconnected from GND
+  digitalWrite(safetyPin,HIGH);
+  while (digitalRead(safetyPin)){   
+    Serial.println("safty loop");
+    delay(5);
+  }
+  for(byte j=0; j<outputPins_count; j++){    //declaring the outputpins 
+    pinMode(outputPins[j],OUTPUT);
+    digitalWrite(outputPins[j],HIGH);  
+  }
+  for(byte j=0; j<inputPins_count; j++){     //declaring the inputpins
+    pinMode(inputPins[j],INPUT_PULLUP);
+  }
+
+  pinMode(AInput, INPUT_PULLUP); // encoder
+  pinMode(BInput, INPUT_PULLUP);
+
+
+  pinMode(NRESET,OUTPUT);       //ADNS5050
+  digitalWrite(NRESET,HIGH);
+  mouse.begin();             //Initialize the ADNS5050
+  delay(100);
+  mouse.sync();    //A sync is performed to make sure the ADNS5050 is communicating
+
+  Keyboard.begin();
+
+  String settings = ReadEEPROMSettings();
+  if(settings.length() == 0) {
+    while (Serial.available() == 0)  {  Serial.println(catVersion);   delay(10);  }
+  } else {
+    parseSettings(settings);
+  }
+}
+
+
+void loop() {
+  ReadSerialcomm();
+
+  ButtonRun();
+}
+
+
+
+void ReadSerialcomm(){
+  if (Serial.available() > 0){
+      String settings = Serial.readStringUntil(strEnder);
+      delay(1000);
+      parseSettings(settings);
+      writeSettingsToEEPROM(settings);
+  }
+}
+
+
+String ReadEEPROMSettings()
+{
+  if(EEPROM.read(0x00) != SETTINGS_VERSION) {
+    return "";
+  }
+  
+  int newStrLen = EEPROM.read(1);
+  char data[newStrLen + 1];
+  for (int i = 0; i < newStrLen; i++)
+  {
+    data[i] = EEPROM.read(2 + i);
+  }
+  data[newStrLen] = '\0';
+  return data;
+}
+
+void writeSettingsToEEPROM(String settings)
+{
+  byte len = settings.length();
+  EEPROM.write(0, SETTINGS_VERSION); // Magic byte, if settings format changes change this.
+  EEPROM.write(1, len);
+  for (int i = 0; i < len; i++)
+  {
+    EEPROM.write(2 + i, settings[i]);
+  }
+}
+
+void parseSettings(String settings) {
+  int inBL = settings.length();
+  int b = 0;
+  int a = 0;
+  byte eIndex = 0;
+  for(b; b < inBL; b++) {  
+    if (settings[b] == delimiter){
+      eventSet[eIndex] = settings.substring(a,b);
+      a = b+1;
+      eIndex++;                    
+      }
+   }
+}
+
+void ButtonRun(){  
+  for (o=0; o<outputPins_count; o++){                  //looping through Outputpins and setting one at a time to LOW 
+      digitalWrite(outputPins[o],LOW);   
+      delayMicroseconds(0);
```

Comment:
I found the buttons still worked without the delay here but didn't do a lot of testing so happy to add it back.